SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+/// ```
+/// %base1 = ref_element_addr %ref
+/// %base2 = alloc_stack $S
+/// %base3 = global_addr @gaddr
+/// %addr1 = struct_element_addr %base1
+/// %access1 = store %v1 to [trivial] %addr1   // accessed address is offset from base
+/// %access2 = store %v2 to [trivial] %base2   // accessed address is base itself
+/// ```
+/// The base address is never inside an access scope.
+struct AccessBase : CustomStringConvertible {
+  let baseAddress: Value
+  let baseKind: AccessBaseKind
+
+  init?(baseAddress: Value) {
+    switch baseAddress {
+    case is RefElementAddrInst   : baseKind = .class
+    case is RefTailAddrInst      : baseKind = .tail
+    case is ProjectBoxInst       : baseKind = .box
+    case is AllocStackInst       : baseKind = .stack
+    case is FunctionArgument     : baseKind = .argument
+    case is GlobalAddrInst       : baseKind = .global
+    default:
+      if baseAddress.definingInstruction is BeginApplyInst &&
+         baseAddress.type.isAddress {
+        baseKind = .yield
+      } else {
+        return nil
+      }
+    }
+
+    self.baseAddress = baseAddress
+  }
+
+  init(baseAddress: Value, baseKind: AccessBaseKind) {
+    self.baseAddress = baseAddress
+    self.baseKind = baseKind
+  }
+
+  var description: String {
+    "\(baseKind) - \(baseAddress)"
+  }
+
+  var isObjectAccess: Bool {
+    return baseKind == .class || baseKind == .tail
+  }
+
+  // If this `AccessBase` is obtained froma a reference type,
+  // results in the reference `ref` and a `component` describing
+  // the projection type (tail or class field) and the corresponding field index
+  var referenceWithPathComponent: (ref: Value, component: (SmallProjectionPath.FieldKind, Int))? {
+    switch baseAddress {
+    case let rea as RefElementAddrInst:
+      return (rea.operand, (.classField, rea.fieldIndex))
+    case let pb as ProjectBoxInst:
+      return (pb.operand, (.classField, pb.fieldIndex))
+    case let rta as RefTailAddrInst:
+      return (rta.operand,(.tailElements, 0))
+    default:
+      return nil
+    }
+  }
+
+  var reference: Value? { referenceWithPathComponent?.ref }
+
+  var isLet: Bool {
+    switch baseAddress {
+    case let rea as RefElementAddrInst:
+      return rea.fieldIsLet
+    case let ga as GlobalAddrInst:
+      return ga.global.isLet
+    default:
+      return false
+    }
+  }
+
+  /// Returns `true` if the address is immediately produced by a stack or box allocation
+  var isLocal: Bool {
+    switch baseKind {
+    case .box:
+      // The operand of the projection can be an argument, in which
+      // case it wouldn't be local
+      return (baseAddress as! ProjectBoxInst).operand is AllocBoxInst
+    case .stack:
+      return true
+    default:
+      return false
+    }
+  }
+
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`.
+  /// When comparing two uniquely identified access bases, it can be followed that if they are not equal,
```

Comment:
"When comparing two uniquely identified access bases and they not equal, then it follows that also the accessed memory addresses do not alias."